### ISSUE TITLE
Bug: <Parsing error: '>' expected.>

### ISSUE THREAD

**Issue by johnboulder** [typescript-eslint/typescript-eslint#5272]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Repro Code

```typescript
import { ESLintUtils } from '@typescript-eslint/utils';
import noRelativePathAnchors from "../no-relative-path-anchors";

const ruleTester = new ESLintUtils.RuleTester({
  parser: '@typescript-eslint/parser',
});

ruleTester.run('no-relative-path-anchors', noRelativePathAnchors, {
  valid: [
    {
      code: '<Link href="https://www.google.com/house">Google</Link>',
    },
  ],
  invalid: [
    {
      code: '<a href="/home">Home</a>',
      errors: [{ messageId: 'noRelativePathAnchor' }],
    },
  ],
});
```

### ESLint Config

```javascript
extends: ../../.eslintrc.yml
parser: 'babel-eslint'
rules:
  no-console: 'off'
  import/no-extraneous-dependencies: 'off'
  react/require-default-props: 'off'
  no-underscore-dangle: 'off'
```

### tsconfig

```jsonc
{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "src",
    "strict": false
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "dist", "**/__tests__/*.spec.ts"]
}
```

### Expected Result

Running this test against my rule should, at the very least, run without errors. 

### Actual Result

1) no-relative-path-anchors
       valid
         <Link href="https://www.google.com/house">Google</Link>:

      AssertionError [ERR_ASSERTION]: A fatal parsing error occurred: Parsing error: '>' expected.
      + expected - actual

      -false
      +true
      
      at runRuleForItem (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:651:13)
      at testValidTemplate (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:704:28)
      at Context.<anonymous> (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:986:29)
      at processImmediate (internal/timers.js:461:21)

  2) no-relative-path-anchors
       invalid
         <a href="/home">Home</a>:

      AssertionError [ERR_ASSERTION]: A fatal parsing error occurred: Parsing error: '>' expected.
      + expected - actual

      -false
      +true
      
      at runRuleForItem (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:651:13)
      at testInvalidTemplate (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:758:28)
      at Context.<anonymous> (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:997:29)
      at processImmediate (internal/timers.js:461:21)

### Additional Info

The custom rule I wrote that I'm struggling with. Perhaps the issue is within this: 
```ts
// @ts-nocheck
import { ESLintUtils, TSESTree } from '@typescript-eslint/utils';

const HREF_ATTRIBUTE_NAME = 'href';
const ANCHOR_ELEMENT_NAME = 'a';

export const noRelativePathAnchors = ESLintUtils.RuleCreator.withoutDocs({
  create(context) {
    const isRelativePathString = (href: string): boolean => /^\/[A-z0-9-._~!$&'()*+;=:@%/]*$/g.test(href);

    return {
      JSXOpeningElement(node: TSESTree.JSXOpeningElement): void {
        if (node.name.name === ANCHOR_ELEMENT_NAME) {
          node.attributes.forEach((attribute) => {
            if (attribute.name === HREF_ATTRIBUTE_NAME) {
              if (isRelativePathString(attribute.value)) {
                context.report({
                  node,
                  messageId: 'noRelativePathAnchor',
                });
              }
            }
          });
        }
      },
    };
  },
  meta: {
    docs: {
      description: 'Disallow use of relative path links and anchors in Single Page App. ',
      recommended: false,
    },
    messages: {
      noRelativePathAnchor:
        'Use of relative path links and anchors circumvents the desired behavior of a single page app. Use `<UitkLink>` instead. ',
    },
    type: 'suggestion',
    schema: [],
  },
  defaultOptions: [],
});

export default noRelativePathAnchors;

```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `5.30.0` |
| `@typescript-eslint/utils`             | `5.30.0` |
| `TypeScript`                           | `4.7.4` |
| `ESLint`                               | `8.18.0` |
| `node`                                 | `12.22.1` |

**Comment by bradzacher** [comment]
set `parserOptions.ecmaFeatures.jsx: true` on your test

```js
{
  code: ' ... ',
  parserOptions: { ecmaFeatures: { jsx: true } },
}
```

This is mentioned at the end of our docs on testing (https://typescript-eslint.io/docs/development/custom-rules#testing-typed-rules)

**Comment by johnboulder** [comment]
Thank you @bradzacher that helped. Is there a guide anywhere that outlines developing rules around jsx elements specifically with `typescript-eslint`, I've struggled to find anything pointing me in the right direction.

**Comment by johnboulder** [comment]
If you look at the rule I shared above, I had to set `// @ts-nocheck` at the top because I'm getting errors with the type checking saying stuff like this: 
<img width="846" alt="Screen Shot 2022-06-29 at 5 45 09 PM" src="https://user-images.githubusercontent.com/9061140/176558577-48dc66a1-982c-45f7-83b7-248b877f44ad.png">

Additionally, the impression I'm getting from running the test against this rule is that `JSXOpeningElement` doesn't ever get called, which is unexpected. I don't see why that would be unless I'm missing some configuration somewhere else.

**Comment by johnboulder** [comment]
The results of the test after I run it: 
```ts
1) no-relative-path-anchors
       invalid
         <a href="/home">Home</a>:

      AssertionError [ERR_ASSERTION]: Should have 1 error but had 0: []
      + expected - actual

      -0
      +1
      
      at testInvalidTemplate (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:773:24)
      at Context.<anonymous> (/Users/jstockwell/IdeaProjects/eg-console-sdk/node_modules/eslint/lib/rule-tester/rule-tester.js:997:29)
      at processImmediate (internal/timers.js:461:21)
```

**Comment by bradzacher** [comment]
TypeScript is correct - the issue is that you are duck-typing your nodes.
You need to refine based on the `type` on a node when there is a union of different nodes possible.

In this instance - `JSXOpeningElement.attributes` has the type `(JSXAttribute | JSXSpreadAttribute)[]` - i.e. your `attribute` variable is of type `JSXAttribute | JSXSpreadAttribute`. `JSXSpreadAttribute` does not have the property `name` - which is what is indicated by the error.

Here's an example of a spread attribute: `<a {...object} />`

This advice doens't just apply to JSX nodes - it applies to the entire AST. If there is a union you must refine the type correctly using `.type`
